### ISSUE TITLE
Product: Rust's `Result` in GraphQL APIs

### ISSUE THREAD

**Issue by JonahPlusPlus** [async-graphql/async-graphql#1649]
## Description of the feature

I've been working on creating a type for handling expected errors in GraphQL (essentially, handling them in the "data" object of the response, instead of the "errors" object).
I wanted to see if this would be appropriate to upstream into the library.

### How it works

This introduces a new enum type called `Product<T, E, F>`, which is similar to `Result<T, E>`, except that it has an additional variant called `Fail(F)`, which encapsulates unexpected errors, which should be placed in "errors".
Where with `Result<T, E>`, `T` maps to the value and `E` maps to an entry in "errors", `Product<T, E, F>` would map `T` and `E` to object fields `ok` and `err`.

It would have to either generate a name from `T` and `E` (like `<T><E>Result`), or use a generic with a trait that provides an associated constant/method with the name (the former is simpler, but I suppose it's possible to make a default marker struct that provides it).

From the client side, the developer can check if `ok` is null, and if so, check `err` for the error.

### Prior art

Before I started on this, I was using a new type `ApiResult<T, E>`, that wrapped `Result<T, E>` and used a macro to implement the API for it.
This become very verbose, since I would have something like `Result<GetUserResult, DatabaseError>` (where `GetUserResult` is a type alias of `ApiResult<User, UserError>`) and had to write `Ok(ApiResult(Ok(user)))` when returning.
With `Product<User, UserError, DatabaseError>`, I could just write `Product::Ok(user)`.

### Limitations

This is very opinionated and I haven't seen this in idiomatic GraphQL queries (that said, I'm rather new to GraphQL).
Still, I think it might be appropriate for a Rust crate for GraphQL to have Rust-oriented utilities.

Another problem is whether or not it is appropriate to add when the `try_trait_v2` feature is not stable.
Using `FromResidual` would be nice to add so converting `Result<T, F>` to `Product<T, E, F>` would be as simple as using `?`.
But this could be resolved for now by adding a "nightly" feature flag for the crate.

## Code example

The definition of `Product<T, E, F>`:
```rust
pub enum Product<T, E, F> {
    Ok(T),
    Err(E),
    Fail(F),
}
```

Exposed API for clients (using a rename strategy) `Product<User, UserError, DatabaseError>`:
```graphql
type UserResult {
  err: UserError
  ok: User
}
```

Possible approaches for naming `Product` in GraphQL:
- Simple concat of T + E:
`Product<User, UserError, Failure>` -> `UserUserErrorResult` 🤮
- Simple concat of T:
`Product<User, UserError, Failure>` -> `UserResult` 🤔 better, but prone to collisions
- Associated constant on trait:
```rust
struct UserResultTy;

impl ProductName for UserResultTy {
  const NAME: &str = "UserResult";
}

type UserResult = Product<User, UserError, DatabaseError, UserResultTy>;
```
- Trait approach with attribute macro:
```rust
#[async_graphql::Product]
type UserResult = Product<User, UserError, DatabaseError>; // expands to the previous example
```
- Alternative macro syntax:
```rust
async_graphql::product!(UserResult<User, UserError, DatabaseError>); // doesn't fit with async_graphql's use of attribute macros IMO
```

Rough example of implementation with renaming strategy: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=eac235f2d167dd4ac639f1d54c50a326

## Note

This could also be implemented as a separate crate.

**Comment by meclondrej** [comment]
How did you make your `ApiResult` a graphql type? I am currently trying write it as you described in "Prior art". I chose a discriminated union to express it, turns out graphql has those (`async_graphql::Union`, they tag the enum variant in `__typename` and include the inner data of the variant), but I can't get it to work with generics.

This code...

```rust
#[derive(Union)]
enum ApiResult<T, E> {
    Ok(T),
    Err(E),
    // i am not including the Fail variant out of simplicity
}
```

...fails to compile because the generics don't necessarily have to be graphql types. Putting a `ContainerType` or an `OutputType` constraint on the generics doesn't help.

**Comment by JonahPlusPlus** [comment]
@meclondrej I just defined it as:
```rust
#[derive(Clone, Copy, From)]
pub struct ApiResult<T, E>(pub Result<T, E>);

impl<T, E> Deref for ApiResult<T, E> {
    type Target = Result<T, E>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
```

Then to use it, I would use a type alias and use an `Object` macro. e.g.
```rust
pub type GetServerResult = ApiResult<Server, ServerError>;

#[async_graphql::Object]
impl GetServerResult {
    async fn success(&self) -> Option<&Server> {
        self.as_ref().ok()
    }

    async fn error(&self) -> Option<ServerError> {
        self.as_ref().err().cloned()
    }
}
```

So, not really nice to use, since you need to write a lot of boilerplate (hence the `Product` type).

**Comment by meclondrej** [comment]
@JonahPlusPlus My take could be wrong on this, but isn't this really a generics issue? GraphQL already has unions that can represent a `Result` (or any other enum) and are supported by `async-graphql`. Making your own way of representing enums (`impl GetServerResult`) wouldn't be necessary if the `Union` derive macro could handle generics. Then, `ApiResult` could be typed universally, with no boilerplate. (by this, i don't mean to say that `Product` is a bad idea and shouldn't be upstreamed)

**Comment by meclondrej** [comment]
I wrote an `ApiResult` implementation without any macros using GraphQL unions, it could help.

<details>
<summary>The code</summary>

```rust
#[derive(async_graphql::SimpleObject)]
#[graphql(name = "Ok")]
struct OkWrapper<T: OutputType> {
    value: T,
}

#[derive(async_graphql::SimpleObject)]
#[graphql(name = "Err")]
struct ErrWrapper<E: OutputType> {
    error: E,
}

#[derive(async_graphql::Union)]
enum ApiResult<T: OutputType, E: OutputType> {
    // the wrappers are here because not all types implement ContainerType (like enums)
    // and because __typename is now locked to "Ok"/"Err".
    Ok(OkWrapper<T>),
    Err(ErrWrapper<E>),
}

impl<T: OutputType, E: OutputType> From<Result<T, E>> for ApiResult<T, E> {
    fn from(value: Result<T, E>) -> Self {
        match value {
            Ok(x) => Self::Ok(OkWrapper { value: x }),
            Err(x) => Self::Err(ErrWrapper { error: x }),
        }
    }
}

impl<T: OutputType, E: OutputType> Into<Result<T, E>> for ApiResult<T, E> {
    fn into(self) -> Result<T, E> {
        match self {
            Self::Ok(wrap) => Ok(wrap.value),
            Self::Err(wrap) => Err(wrap.error),
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq, async_graphql::Enum)]
enum TestError {
    FailureA,
    FailureB,
}

type TestResult = ApiResult<String, TestError>;

struct QueryRoot;

#[Object]
impl QueryRoot {
    async fn test(&self, n: i32) -> TestResult {
        if n == 1 {
            return Err(TestError::FailureA).into();
        } else if n == 2 {
            return Err(TestError::FailureB).into();
        }
        Ok("yay!".to_string()).into()
    }
}
```
</details>

**Comment by JonahPlusPlus** [comment]
Yeah, if the union macro supported generics, then it would work, but I don't think it will (unless some sort of naming strategy trait like `ProductName` as I described above is implemented).

It's the same for your snippet (AFAIK), the `OkWrapper` and `ErrWrapper` need `#[graphql(concrete(name = "SomeName", params(SomeType)))]` to change the name, but that isn't ergonomic for large codebases.

This all comes down to GraphQL not supporting generic types.

**Comment by meclondrej** [comment]
Why would the wrappers need to change names?

**Comment by JonahPlusPlus** [comment]
Because names can't conflict in GraphQL (AFAIK) and types can't be generic. So if your API uses `Result<Bar, Foo>` and `Result<Biz, Foo>`, there would need to be a way to convert them to concrete types like `ResultBarFoo` and `ResultBizFoo` (and in this case, then the wrappers like `OkBar`, `OkBiz` and `ErrFoo`).

**Comment by meclondrej** [comment]
Better approach using macros (inspired by [this](https://github.com/async-graphql/async-graphql/issues/1633#issuecomment-2559834139)), uses the paste crate for ident concating:

[The code](https://gist.github.com/meclondrej/a5e356f48f3e622272f2b90c1aa701a0)

This creates custom wrapper names by just concating Ok or Err to the result's name.
